restore-hooks_product-requirements.md

```diff
@@ -0,0 +1,376 @@
+
+**Note: **Please refer to the Requirements section of this document for more details on what items are P0 (must have in the MVP), P1 (should not ship without for the MVP), P2 (nice to haves).
+
+**<span style="text-decoration:underline;">USE CASE 1</span>**
+
+
+    **Title:** Run restore hook before pod restart.
+
+
+    **Description:  **As a user, I would like to run a restore hook before the applications in my pod are restarted.
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 2</span>**
+
+
+    **Title: **Allow restore hook to run on non-kubernetes databases
+
+
+    **Description: **As a user, I would like to run restore hook operations even on databases that are external to kubernetes (such as postgres, elastic, etc…).
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 3</span>**
+
+
+    **Title: **Run restore at pod level.
+
+
+    **Description: **As a user, I would like to make sure that I can run the restore hook at the pod level. And, I would like the option to run this with an annotation flag in line or using an init container.**<span style="text-decoration:underline;"> </span>**The restore pre-hook should allow the user to run the command on the container where the pre-hook should be executed. Similar to the backup hooks, this hook should run to default to fun on the first container in the pod. 
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 4</span>**
+
+
+    **Title:** Specify container in which to run pre-hook
+
+
+    **Description: **As a user, if I do not want to run the pre-hook command on the first container in the pod (default container), I would like the option to annotate the specific container that i would like the hook to run in.
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 5</span>**
+
+
+    **Title: **Check for latest snapshot 
+
+
+    **Description: **As a user, I would like Velero to run a check for the latest snapshot in object storage prior to starting restore operations on a pod. 
+
+**<span style="text-decoration:underline;">______________________________________________________________</span>**
+
+**<span style="text-decoration:underline;">USE CASE 6</span>**
+
+
+    **Title: **Specify backup snapshot to use for restore
+
+
+    **Description: **As a user, I would like the option to specify a specific backup snapshot for use by Velero during restore create. I would like to do this instead of using the default most recent backup restore snapshot.
```

Comment:
Can you expand on this use case a little more? This sounds like how `velero create restore --from-backup <backup name>` works already today, but I may be missing something.

`velero create restore --from-schedule` will create a restore from the latest backup associated with a given schedule, but the `--from-backup` version allows a user to choose a specific backup, so long as it still exists.

Reply:
this doesn't sound like a use case for restore hooks but rather for restore create

Reply:
The intent of this was that if a user preferred to specify a snapshot they would like to use other than the default 'latest' snapshot that was taken from the velero create restore --from-schedule, that the user could instead specify a snapshot they would like to use instead.

This is most applicable in security use cases or corruption use cases where, the data contained in the latest snapshot may have been corrupt. Thoughts on this @nrb @yaronp68 ?

Reply:
I think that's a valid use case, but I do believe `velero create restore --from-backup` meets it already, as `--from-backup` allows the user to select any valid Velero backup that exist.

Backups associated with a Schedule all exist in the Kubernetes API server with a label on them, and the `--from-schedule` option grabs all Backups labeled with the Schedule's name, sorts them by date created, then grabs the latest one. So Backups created from a Schedule aren't particularly special other than the label once created.

Would you agree that this meets the use case?